service/controller/accessors/cert.go

```diff
@@ -1,4 +1,4 @@
-package key
+package accessors
```

Comment:
Happy to hear suggestions for a better name, but this seemed fairly descriptive to me.

Reply:
Why shouldn't this remain `key`?

Reply:
There's already a `key` package there and it looks like something unrelated to me. In fact I'm not sure why the other one is called `key` either. Anyway, I'm just not too familiar with this, do you think they should be merged?

Reply:
I'd prefer we keep the key name. As to me its both clear and short. We use the key package whenever we need to read values from custom resources.

For the clusterapi controller we should only need a single key package. For legacy we will still need key packages per provider.

Reply:
I think they should be merged as well.

Reply:
OK, so the problem is there's a lot of functions with the same names in `service/controller/clusterapi/v22/key` and in `service/controller/key` but they do things differently e.g. the clusterapi version uses a `LabelsGetter` interface while the other version just gets data injected directly, e.g.:

``` go
// from `service/controller/key/key.go`
func CertConfigName(clusterID string, cert certs.Cert) string {
	return fmt.Sprintf("%s-%s", clusterID, cert)
}

// from `service/controller/clusterapi/v22/key/cert.go`
func CertConfigName(getter LabelsGetter, name string) string {
	return fmt.Sprintf("%s-%s", ClusterID(getter), name)
}
```

or

``` go
// from `service/controller/key/key.go`
func ClusterID(clusterGuestConfig v1alpha1.ClusterGuestConfig) string {
	return clusterGuestConfig.ID
}

// from `service/controller/clusterapi/v22/key/common.go`
func ClusterID(getter LabelsGetter) string {
	return getter.GetLabels()[label.Cluster]
}
```

any ideas how to reconcile those differences? Pick one and rewrite call sites using the other?

Reply:
Anything referencing `v1alpha1.ClusterGuestConfig` should be removed. The clusterapi controller only watches `v1alpha1.Cluster` from `sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1`.

Reply:
The only thing you should need from the provider key packages like `controller/aws/key` are the app and chart structs.

https://github.com/giantswarm/cluster-operator/blob/24f07f0eef042b4a993a480e20c39690756927f8/service/controller/aws/key/key.go#L11-L51
https://github.com/giantswarm/cluster-operator/blob/24f07f0eef042b4a993a480e20c39690756927f8/service/controller/aws/key/key.go#L53-L69

I would keep these for now and we can decouple them in a later PR.

Reply:
I think the above can be also moved with appropriate suffixes/prefixes. Like `AWSAppSepcs` or `AppSpecsAWS`. Possibly in a separate PR as @rossf7 suggests.

Reply:
Thanks for the pointers BTW, re-working this now and will probably split into two PRs cause it's large-ish.

Another question, the `ChartSpecs` functions don't seem to be used anywhere, am I missing something? Are you sure we want to keep them?

Reply:
Delete and try to compile :D

Reply:
> Another question, the ChartSpecs functions don't seem to be used anywhere, am I missing something? Are you sure we want to keep them?

Good point. ChartSpecs can be removed. In legacy they are used to migrate from legacy chartconfig CRs to new shiny app CRs.

For clusterapi controller we just create app CRs and don't worry about the legacy shite.

Reply:
Fantastic, will do that.

OK I did some branch surgery to remove that commit completely and am working on and will push a separate PR dealing with all `key` packages including merging of the different providers. In the meantime this can go ahead as a separate thing which will make it easier to review the `key` changes later.